src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/validation/FlowValidationFsm.java

```diff
@@ -116,47 +115,58 @@ protected void receiveData(State from, State to,
         }
 
-        List<SwitchId> switchIds = service.getSwitchIdListByFlowId(flowId);
+        switchIdSet.addAll(service.getSwitchIdListByFlowId(flowId));
 
-        awaitingRules = switchIds.size();
+        awaitingRules = switchIdSet.size();
         log.debug("Send commands to get rules on the switches");
-        switchIds.forEach(switchId ->
+        switchIdSet.forEach(switchId ->
                 getCarrier().sendSpeakerRequest(flowId, new DumpRulesForFlowHsRequest(switchId)));
 
-        log.debug("Send commands to get meters on the termination switches");
-        awaitingMeters = TERMINATION_SWITCHES_COUNT;
-        getCarrier().sendSpeakerRequest(flowId, new DumpMetersForFlowHsRequest(flow.getSrcSwitchId()));
-        getCarrier().sendSpeakerRequest(flowId, new DumpMetersForFlowHsRequest(flow.getDestSwitchId()));
-
-        log.debug("Send commands to get groups on the termination switches");
-        awaitingGroups = TERMINATION_SWITCHES_COUNT;
-        getCarrier().sendSpeakerRequest(flowId, new DumpGroupsForFlowHsRequest(flow.getSrcSwitchId()));
-        getCarrier().sendSpeakerRequest(flowId, new DumpGroupsForFlowHsRequest(flow.getDestSwitchId()));
+        awaitingMeters = 1;
+        awaitingGroups = 1;
+        if (!flow.isOneSwitchFlow()) { // avoid duplicates for single sw rules
+            awaitingMeters++;
+            log.debug("Send commands to get meters on the termination switches");
+            getCarrier().sendSpeakerRequest(flowId, new DumpMetersForFlowHsRequest(flow.getSrcSwitchId()));
+            getCarrier().sendSpeakerRequest(flowId, new DumpMetersForFlowHsRequest(flow.getDestSwitchId()));
+            awaitingGroups++;
+            log.debug("Send commands to get groups on the termination switches");
+            getCarrier().sendSpeakerRequest(flowId, new DumpGroupsForFlowHsRequest(flow.getSrcSwitchId()));
+            getCarrier().sendSpeakerRequest(flowId, new DumpGroupsForFlowHsRequest(flow.getDestSwitchId()));
+        } else {
+            log.debug("Send commands to get meters on the termination switches");
+            getCarrier().sendSpeakerRequest(flowId, new DumpMetersForFlowHsRequest(flow.getSrcSwitchId()));
+            log.debug("Send commands to get groups on the termination switches");
+            getCarrier().sendSpeakerRequest(flowId, new DumpGroupsForFlowHsRequest(flow.getSrcSwitchId()));
+        }
     }
 
     protected void receivedRules(State from, State to,
                                  Event event, Object context) {
-        SwitchFlowEntries switchFlowEntries = (SwitchFlowEntries) context;
-        log.info("Switch rules received for switch {}", switchFlowEntries.getSwitchId());
-        receivedRules.add(switchFlowEntries);
+        FlowDumpResponse flowDumpResponse = (FlowDumpResponse) context;
+        log.info("Switch rules received for switch {}", Optional.ofNullable(flowDumpResponse.getSwitchId())
+                .map(SwitchId::toString).orElse("No switch Id"));
```

Comment:
I think you can keep here the old way: 
`log.info("Switch rules received for switch {}", flowDumpResponse.getSwitchId());`

There is not much different between messages 
`Switch rules received for switch null` and 
`Switch rules received for switch No switch Id`

Reply:
thanks, fixed